### PR TITLE
Add params collections

### LINKED ISSUE #40005: [C# 13]: New Feature - Params collections
### Visual Studio release

17.10p3

### Link to speclet

- [Speclet](https://github.com/dotnet/csharplang/blob/main/proposals/params-collections.md)

### Link to Pull request

- [PR](https://github.com/dotnet/roslyn/pulls/72511)

### New behavior

This feature relaxes restrictions on the type of parameters declared with the `params` modifier. Other collection types are now allowed:  `Span<T>`, `ReadOnlySpan<T>`, `IEnumerable<T>`, and others. 

### Areas affected

- [X] **Add to What's new in C# 13**: Add links to updated docs in the What's new in C# 13 article.
- [X] **Update language reference**: Detail the areas in the language reference that will be updated for this feature.
- [X] **Update compiler messages**: Add new (or updated) pages for new compiler messages. Include related message.
- [X] *Include in *Everyday C#* section**: Is this new feature included in Everyday C#?
- [ ] **Tutorials**: Are new tutorials required for this feature? If so, detail below

### Language reference tasks

The page on the `params` keyword should be updated to include other items.

### Tutorial tasks

_No response_

### Everyday C# tasks

Add the new `params` capability when discussing `params` and the allowed signatures for `Main`. (Check the spec to see if Main is affected. Either way, we should make clear what the behavior is).

### LINKED ISSUE #40796: Confusing C# 13 example
### Describe the issue or suggestion

I find the example for implicit index example to be confusing: https://github.com/dotnet/docs/blob/6ae8f6122a9cc8c23d1e4e4d208e06d023123961/docs/csharp/whats-new/csharp-13.md#L47.

The primary issue is that I have to guess what `v` and `S` are. I realize that they are not important for the example, however, they are the first bits of code that I have to look at and parse. Also, is this code that one would ever write? Maybe this feature is just too niche to get this much fanfare?

I like these pattern matching examples: https://learn.microsoft.com/dotnet/csharp/fundamentals/functional/pattern-matching.

---
#### Document Details

⚠ *Do not edit this section. It is required for learn.microsoft.com ➟ GitHub issue linking.*

* ID: aac84693-3ecb-f53b-3841-eae6c5aeb653
* Version Independent ID: 51d7b1c3-f822-0c65-da31-47a3fcd22dd3
* Content: [What's new in C# 13 - C# Guide - C#](https://learn.microsoft.com/en-us/dotnet/csharp/whats-new/csharp-13)
* Content Source: [docs/csharp/whats-new/csharp-13.md](https://github.com/dotnet/docs/blob/main/docs/csharp/whats-new/csharp-13.md)
* Service: **dotnet-csharp**
* Sub-service: **whats-new**
* GitHub Login: @BillWagner
* Microsoft Alias: **wiwagn**

---
[Associated WorkItem - 256379](https://dev.azure.com/msft-skilling/Content/_workitems/edit/256379)

### PR DESCRIPTION
**Pull request by BillWagner**
Update descriptions for the `params` modifier to allow collections.

Contributes to #40005

This might be easiest to review commit by commit:

1. **Update What's new, language reference, and programming guide where `params` is mentioned.** Search the content for `params` and make updates as needed.
1. **Review articles and samples with smaller scope changes.** A general edit pass.
1. **Review methods.md**: Major edit pass, move samples and trim unused samples. Small updates to the sample code.
1. **Review programming-guide Object Initializers article**: Major edit pass, move samples. Small updates to the samples.
1. **Fix warnings from related samples moving**.
1. **Bug fix**:  Fixes #40796

---

#### Internal previews

| 📄 File | 🔗 Preview link |
|:--|:--|
| [docs/csharp/language-reference/keywords/method-parameters.md](https://github.com/dotnet/docs/blob/96b0062cf613bf7e5b9d2d2b63d2c937c1da84f0/docs/csharp/language-reference/keywords/method-parameters.md) | [Method Parameters](https://review.learn.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/method-parameters?branch=pr-en-us-40978) |
| [docs/csharp/language-reference/operators/lambda-expressions.md](https://github.com/dotnet/docs/blob/96b0062cf613bf7e5b9d2d2b63d2c937c1da84f0/docs/csharp/language-reference/operators/lambda-expressions.md) | ["Lambda expressions - Lambda expressions and anonymous functions"](https://review.learn.microsoft.com/en-us/dotnet/csharp/language-reference/operators/lambda-expressions?branch=pr-en-us-40978) |
| [docs/csharp/methods.md](https://github.com/dotnet/docs/blob/96b0062cf613bf7e5b9d2d2b63d2c937c1da84f0/docs/csharp/methods.md) | [Methods in C\#](https://review.learn.microsoft.com/en-us/dotnet/csharp/methods?branch=pr-en-us-40978) |
| [docs/csharp/misc/cs0225.md](https://github.com/dotnet/docs/blob/96b0062cf613bf7e5b9d2d2b63d2c937c1da84f0/docs/csharp/misc/cs0225.md) | [Compiler Error CS0225](https://review.learn.microsoft.com/en-us/dotnet/csharp/misc/cs0225?branch=pr-en-us-40978) |
| [docs/csharp/programming-guide/classes-and-structs/how-to-initialize-a-dictionary-with-a-collection-initializer.md](https://github.com/dotnet/docs/blob/96b0062cf613bf7e5b9d2d2b63d2c937c1da84f0/docs/csharp/programming-guide/classes-and-structs/how-to-initialize-a-dictionary-with-a-collection-initializer.md) | [How to initialize a dictionary with a collection initializer (C# Programming Guide)](https://review.learn.microsoft.com/en-us/dotnet/csharp/programming-guide/classes-and-structs/how-to-initialize-a-dictionary-with-a-collection-initializer?branch=pr-en-us-40978) |
| [docs/csharp/programming-guide/classes-and-structs/how-to-initialize-objects-by-using-an-object-initializer.md](https://github.com/dotnet/docs/blob/96b0062cf613bf7e5b9d2d2b63d2c937c1da84f0/docs/csharp/programming-guide/classes-and-structs/how-to-initialize-objects-by-using-an-object-initializer.md) | [docs/csharp/programming-guide/classes-and-structs/how-to-initialize-objects-by-using-an-object-initializer](https://review.learn.microsoft.com/en-us/dotnet/csharp/programming-guide/classes-and-structs/how-to-initialize-objects-by-using-an-object-initializer?branch=pr-en-us-40978) |
| [docs/csharp/programming-guide/classes-and-structs/object-and-collection-initializers.md](https://github.com/dotnet/docs/blob/96b0062cf613bf7e5b9d2d2b63d2c937c1da84f0/docs/csharp/programming-guide/classes-and-structs/object-and-collection-initializers.md) | [Object and Collection Initializers (C# Programming Guide)](https://review.learn.microsoft.com/en-us/dotnet/csharp/programming-guide/classes-and-structs/object-and-collection-initializers?branch=pr-en-us-40978) |
| [docs/csharp/whats-new/csharp-13.md](https://github.com/dotnet/docs/blob/96b0062cf613bf7e5b9d2d2b63d2c937c1da84f0/docs/csharp/whats-new/csharp-13.md) | [What's new in C# 13](https://review.learn.microsoft.com/en-us/dotnet/csharp/whats-new/csharp-13?branch=pr-en-us-40978) |

### CODE REVIEW

File: docs/csharp/methods.md [hunk 1]
```diff
@@ -32,112 +29,100 @@ These parts together form the method signature.
 
 The following example defines a class named `Motorcycle` that contains five methods:
 
-[!code-csharp[csSnippets.Methods#40](../../samples/snippets/csharp/concepts/methods/methods40.cs#40)]
-
-The `Motorcycle` class includes an overloaded method, `Drive`. Two methods have the same name, but must be differentiated by their parameter types.
+:::code language="csharp" source="snippets/methods/methods40.cs" id="snippet40":::
 
-<a name="invocation"></a>
+The `Motorcycle` class includes an overloaded method, `Drive`. Two methods have the same name, but you must differentiate them by their parameter types.
```
Comment (gewarren) [inline docs/csharp/methods.md:34]:
```suggestion
The `Motorcycle` class includes an overloaded method, `Drive`. Two methods have the same name, but are differentiated by their parameter types.
```

Comment (gewarren) [reply docs/csharp/methods.md:34]: It's describing someone else's code, not code the reader would/could write.

File: docs/csharp/methods.md [hunk 2]
```diff
@@ -32,112 +29,100 @@ These parts together form the method signature.
...
 
-<a name="paramarray"></a>
-
-### Parameter arrays
+### Parameter collections
 
-Sometimes, the requirement that you specify the exact number of arguments to your method is restrictive. By using the `params` keyword to indicate that a parameter is a parameter array, you allow your method to be called with a variable number of arguments. The parameter tagged with the `params` keyword must be an array type, and it must be the last parameter in the method's parameter list.
+Sometimes, the requirement that you specify the exact number of arguments to your method is restrictive. By using the `params` keyword to indicate that a parameter is a parameter collection, you allow your method to be called with a variable number of arguments. The parameter tagged with the `params` keyword must be a collection type, and it must be the last parameter in the method's parameter list.
 
 A caller can then invoke the method in either of four ways:
```
Comment (gewarren) [inline docs/csharp/methods.md:104]:
```suggestion
A caller can then invoke the method in either of four ways for the `params` parameter:
```

File: docs/csharp/misc/cs0225.md [hunk 3]
```diff
@@ -10,9 +10,9 @@ ms.assetid: 0b0cd72b-c47a-44d1-9b27-d1a1fad06807
 ---
 # Compiler Error CS0225
 
-The params parameter must be a single-dimensional array
+The params parameter must be a single-dimensional array or collection
   
- When using the [params](../language-reference/keywords/method-parameters.md#params-modifier) keyword, you must specify a single-dimensional array of the data type. For more information, see [Methods](../programming-guide/classes-and-structs/methods.md).  
+ When using the [params](../language-reference/keywords/method-parameters.md#params-modifier) keyword, you must specify a single-dimensional or collection of the data type. For more information, see [Methods](../programming-guide/classes-and-structs/methods.md).
```
Comment (gewarren) [inline docs/csharp/misc/cs0225.md:15]:
```suggestion
 When using the [params](../language-reference/keywords/method-parameters.md#params-modifier) keyword, you must specify a single-dimensional array or collection of the data type. For more information, see [Methods](../programming-guide/classes-and-structs/methods.md).  
```

File: docs/csharp/programming-guide/classes-and-structs/object-and-collection-initializers.md [hunk 4]
```diff
@@ -125,48 +124,48 @@ Required init-only properties support immutable structures while allowing natura
...
 
-This initializer example calls <xref:System.Collections.Generic.Dictionary%602.Add(%600,%601)> to add the three items into the dictionary. These two different ways to initialize associative collections have slightly different behavior because of the method calls the compiler generates. Both variants work with the `Dictionary` class. Other types may only support one or the other based on their public API.
+This initializer example calls <xref:System.Collections.Generic.Dictionary%602.Add(%600,%601)> to add the three items into the dictionary. These two different ways to initialize associative collections have slightly different behavior because of the method calls the compiler generates. Both variants work with the `Dictionary` class. Other types might only support one or the other based on their public API.
 
 ## Object Initializers with collection read-only property initialization
 
-Some classes may have collection properties where the property is read-only, like the `Cats` property of `CatOwner` in the following case:
+Some classes can have collection properties where the property is read-only, like the `Cats` property of `CatOwner` in the following case:
```
Comment (gewarren) [inline docs/csharp/programming-guide/classes-and-structs/object-and-collection-initializers.md:164]:
```suggestion
Some classes might have collection properties where the property is read-only, like the `Cats` property of `CatOwner` in the following case:
```

File: docs/csharp/whats-new/csharp-13.md [hunk 5]
```diff
@@ -21,11 +22,17 @@ New features are added to the "What's new in C#" page when they're available in
 
 [!INCLUDE [released-version-feedback](./includes/released-feedback.md)]
 
+## `params` collections
+
+The `params` modifier isn't limited to array types. You can now use `param` with any recognized collection type. In addition to arrays, <xref:System.Span%601?displayProperty=nameWithType>, <xref:System.ReadOnlySpan%601?displayProperty=nameWithType>, types that implement <xref:System.Collections.Generic.IEnumerable%601?displayProperty=nameWithType> in addition to having an `Add` method. In addition to concrete types, interfaces <xref:System.Collections.Generic.IEnumerable%601?displayProperty=nameWithType>, <xref:System.Collections.Generic.IReadOnlyCollection%601?displayProperty=nameWithType>, <xref:System.Collections.Generic.IReadOnlyList%601?displayProperty=nameWithType>, <xref:System.Collections.Generic.ICollection%601?displayProperty=nameWithType>, or <xref:System.Collections.Generic.IList%601?displayProperty=nameWithType> can be used.
```
Comment (gewarren) [inline docs/csharp/whats-new/csharp-13.md:27]:
```suggestion
The `params` modifier isn't limited to array types. You can now use `params` with any recognized collection type, including <xref:System.Span%601?displayProperty=nameWithType>, <xref:System.ReadOnlySpan%601?displayProperty=nameWithType>, and types that implement <xref:System.Collections.Generic.IEnumerable%601?displayProperty=nameWithType> and have an `Add` method. In addition to concrete types, the interfaces <xref:System.Collections.Generic.IEnumerable%601?displayProperty=nameWithType>, <xref:System.Collections.Generic.IReadOnlyCollection%601?displayProperty=nameWithType>, <xref:System.Collections.Generic.IReadOnlyList%601?displayProperty=nameWithType>, <xref:System.Collections.Generic.ICollection%601?displayProperty=nameWithType>, and <xref:System.Collections.Generic.IList%601?displayProperty=nameWithType> can also be used.
```

**Review by gewarren — APPROVED** [review]

File: docs/csharp/language-reference/keywords/method-parameters.md [hunk 6]
```diff
@@ -197,16 +197,39 @@ The only method call where the argument is passed by reference is the final one.
...
 
-If the declared type of the `params` parameter isn't a single-dimensional array, compiler error [CS0225](../../misc/cs0225.md) occurs.
+The declared type of the `params` parameter must be a collection type. Recognized collection types are:
+
+- A single dimensional *array type* `T[]`, in which case the *element type* is `T`
+- A *span type*
+  - `System.Span<T>`
+  - `System.ReadOnlySpan<T>`  
+  in which cases the *element type* is `T`
+- A *type* with an accessible *create method* with a corresponding *element type*.
```
[the remark this answers is not in the document]
Comment (BillWagner) [reply docs/csharp/language-reference/keywords/method-parameters.md]: It must follow the same pattern used in collection expressions. I  added a clarifying sentence and a link to the collection builder article.

**Pull request merged**